src/main/java/net/sf/rails/game/GameManager.java

```diff
@@ -1295,8 +1299,63 @@ public void registerCompanyBankruptcy (PublicCompany company) {
         or.finishTurn();
     }
 
+    /**
+     * Process the effects of a player going bankrupt, without ending the game.
+     * This code applies to (most of) the David Hecht games.
+     * So far 1826, 18EU, 18VA, 18Scan have been identified to use this code.
+     */
     protected void processPlayerBankruptcy() {
```

Comment:
I personally don't like the idea to provide a default implementations that is used by a subset of all implemented variants.
This is not really targeted only to this method but to most other classes as well, where we have similar issues, because we implement a default behaviour based on one game variant (normally 1830) and then "change" it in a subvariant.
This makes it quite easy in my opinion to introduce bugs because the "fallback" behaviour, which is called when you don't provide a dedicated behaviour, is always the one from 1830.

Ideally I think we should make the behaviour in the variants more explicit. Maybe we could use a [mixin approach](https://en.wikipedia.org/wiki/Mixin)? In this case be case we could try to define an "empty" interface containing the method stub(s), an "default" interface containing the default implementation(s) and a number of differing implementations in the non-default variants. 

Alternatively we could think about externalizing the default behaviour in utility classes, which we can then explicitly call in all variants.

Reply:
I could not immediately find examples of mixin constructs in Java, so I don't really get what you mean. 

Would your "default" interface implement the "empty" interface and the main (not game-specific) class, such as OperatingRound,  implement the default interface? Wouldn't then the default methods also extend to all games that don't override these? In that case, I don't see the difference with having stubs, or default methods, in the main class.

Or do you intend to restrict such interfaces to games that show a certain behaviour that does not apply to other games?
We could think of bankrupty processing, which does not apply to all games where a bankruptcy ends a game. The other games could then be the only ones implementing such interfaces. This could be a sensible idea.

In Rails 1.x we had, on someone's proposal, additional interfaces for all kinds of objects. Train and TrainI, Tile and TileI and so on and so on. These were hardly used, and thankfully removed in Rails 2.x.

Keeping the explosion of separate Java files under control is also a worthy goal to me.